diag_manager/fms_send_data_statfun.F90

```diff
@@ -0,0 +1,1139 @@
+                        & ofb(is-hi:ie-hi,js-hj:je-hj,:,sample) + &
+                        & weight_procs%fwf_3d_ptr(field(f1:f2,f3:f4,ksr:ker), weight1, pow_value)
+                     ELSE
+!$OMP CRITICAL
+                        ofb(is-hi:ie-hi,js-hj:je-hj,:,sample) =&
+                        & ofb(is-hi:ie-hi,js-hj:je-hj,:,sample) + &
+                        & weight_procs%fwf_3d_ptr(field(f1:f2,f3:f4,ksr:ker), weight1, pow_value)
+!$OMP END CRITICAL
+                     END IF
+
+                  ELSE NDCMP_RKR_4_IF
+                     IF ( debug_diag_manager ) THEN
+                        CALL update_bounds(out_num, is-hi, ie-hi, js-hj, je-hj, ks, ke)
+                        CALL check_out_of_bounds(out_num, diag_field_id, err_msg=err_msg_local)
+                        IF ( err_msg_local /= '' ) THEN
+                           IF (fms_error_handler('diag_manager_mod::send_data_3d', err_msg_local, err_msg) ) THEN
+                              succeded = .FALSE.
+                              RETURN
+                           END IF
+                        END IF
+                     END IF
+                     IF( numthreads > 1 .AND. phys_window ) then
+                        ofb(is-hi:ie-hi,js-hj:je-hj,ks:ke,sample) =&
+                        & ofb(is-hi:ie-hi,js-hj:je-hj,ks:ke,sample) +&
+                        & weight_procs%fwf_3d_ptr(field(f1:f2,f3:f4,ks:ke), weight1, pow_value)
+                     ELSE
+!$OMP CRITICAL
+                        ofb(is-hi:ie-hi,js-hj:je-hj,ks:ke,sample) =&
+                        & ofb(is-hi:ie-hi,js-hj:je-hj,ks:ke,sample) +&
+                        & weight_procs%fwf_3d_ptr(field(f1:f2,f3:f4,ks:ke), weight1, pow_value)
+                        !!
+!$OMP END CRITICAL
+                     END IF
+                  END IF NDCMP_RKR_4_IF
+!$OMP CRITICAL
+                  IF ( .NOT.phys_window ) output_fields(out_num)%count_0d(sample) =&
+                  & output_fields(out_num)%count_0d(sample) + weight1
+!$OMP END CRITICAL
+               END IF MISSVAL_PR_3_IF
+            END IF MASK_PR_2_IF ! if mask present
+         END IF MASK_VAR_IF
+      END ASSOCIATE
+
+!$OMP CRITICAL
+      IF ( .NOT.need_compute .AND. .NOT.reduced_k_range )&
+      & output_fields(out_num)%num_elements(sample) =&
+      & output_fields(out_num)%num_elements(sample) + (ie-is+1)*(je-js+1)*(ke-ks+1)
+      IF ( reduced_k_range ) &
+      & output_fields(out_num)%num_elements(sample) = output_fields(out_num)%num_elements(sample) +&
+      & (ie-is+1)*(je-js+1)*(ker-ksr+1)
+!$OMP END CRITICAL
+
+      succeded = .TRUE.
+      RETURN
+
+   END FUNCTION AVERAGE_THE_FIELD
+
+   FUNCTION SAMPLE_THE_FIELD (diag_field_id, field, out_num, &
+   & mask, sample, missvalue, missvalue_present, &
+   & l_start, l_end, idx_cfg, err_msg,  err_msg_local) result( succeded )
+      INTEGER, INTENT(in) :: diag_field_id
+      REAL, DIMENSION(:,:,:), INTENT(in) :: field
```

Comment:
This will need to support real\*8 and real\*4 and potentially more/less dimensions.